lib/client.js

```diff
@@ -0,0 +1,156 @@
+var WS = require('ws');
+var debug = require('debug')('docker-exec-websocket-server:lib:client');
+var url = require('url');
+var events = require('events');
+var _ = require('lodash');
+var assert = require('assert');
+var msgcode = require('../lib/messagecodes.js');
+var through = require('through');
+
+// Makes a new buffer with code number (ideally 0x00-0xFF)
+// appended to the start of the given buffer
+function muxBuffer (code, buffer) {
+  return Buffer.concat([new Buffer([code]), buffer]);
+}
+
+class DockerExecWebsocketClient extends events.EventEmitter {
+  constructor () {
+    super();
+  }
+
+  /* Makes a client program with unbroken stdin, stdout, stderr streams
+   * is also an EventEmitter with 'exit' event
+   *
+   * Required options:
+   * url parts (hostname port pathname)
+   * or url
+   * tty: whether or not we expect VT100 style output
+   * command: array or string of command to be run in exec
+   * not supported yet: ssl cert
+   */
+  static async createClient (options) {
+    var client = new DockerExecWebsocketClient();
+    options = _.defaults({}, options, {
+      tty: true,
+      command: ['/bin/bash'],
+      protocol: 'ws',
+      slashes: true,
+    });
+    options.query = {
+      tty: encodeURIComponent(options.tty),
+      command: encodeURIComponent(JSON.stringify(options.command)),
```

Comment:
No need for this... check docs for `url.format`. In fact the double encoding here is crazy...